internal/cache/file/downloader/job_manager_parallel_downloads_test.go

```diff
@@ -0,0 +1,199 @@
+		readReqSize            int
+		maxDownloadParallelism int
+		downloadOffset         int64
+		expectedOffset         int64
+		subscribedOffset       int64
+	}{
+		{
+			name:                   "download in chunks of concurrency * readReqSize",
+			objectSize:             15 * util.MiB,
+			readReqSize:            4,
+			maxDownloadParallelism: 3,
+			subscribedOffset:       7,
+			downloadOffset:         10,
+			expectedOffset:         12 * util.MiB,
+		},
+		{
+			name:                   "download only upto the object size",
+			objectSize:             10 * util.MiB,
+			readReqSize:            4,
+			maxDownloadParallelism: 3,
+			subscribedOffset:       7,
+			downloadOffset:         10,
+			expectedOffset:         10 * util.MiB,
+		},
+	}
+	for _, tc := range tbl {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			cache, cacheDir := configureCache(t, 2*tc.objectSize)
+			storageHandle := configureFakeStorage(t)
+			bucket := storageHandle.BucketHandle(storage.TestBucketName, "")
+			minObj := createObjectInStoreAndInitCache(t, cache, bucket, "path/in/gcs/foo.txt", tc.objectSize)
+			jm := NewJobManager(cache, util.DefaultFilePerm, util.DefaultDirPerm, cacheDir, 2, &config.FileCacheConfig{EnableParallelDownloads: true,
+				DownloadParallelismPerFile: math.MaxInt, ReadRequestSizeMB: tc.readReqSize, EnableCrcCheck: true, MaxDownloadParallelism: tc.maxDownloadParallelism})
+			job := jm.CreateJobIfNotExists(&minObj, bucket)
+			subscriberC := job.subscribe(tc.subscribedOffset)
+
+			_, err := job.Download(context.Background(), 10, false)
+
+			timeout := time.After(1 * time.Second)
+			for {
+				select {
+				case jobStatus := <-subscriberC:
+					if assert.Nil(t, err) {
+						assert.Equal(t, tc.expectedOffset, jobStatus.Offset)
+					}
+					return
+				case <-timeout:
+					assert.Fail(t, "Test timed out")
+					return
+				}
+			}
+		})
+	}
+}
+
+func TestMultipleConcurrentDownloads(t *testing.T) {
+	t.Parallel()
+	storageHandle := configureFakeStorage(t)
+	cache, cacheDir := configureCache(t, 30*util.MiB)
+	bucket := storageHandle.BucketHandle(storage.TestBucketName, "")
+	minObj1 := createObjectInStoreAndInitCache(t, cache, bucket, "path/in/gcs/foo.txt", 10*util.MiB)
+	minObj2 := createObjectInStoreAndInitCache(t, cache, bucket, "path/in/gcs/bar.txt", 5*util.MiB)
+	jm := NewJobManager(cache, util.DefaultFilePerm, util.DefaultDirPerm, cacheDir, 2, &config.FileCacheConfig{EnableParallelDownloads: true,
+		DownloadParallelismPerFile: math.MaxInt, ReadRequestSizeMB: 2, EnableCrcCheck: true, MaxDownloadParallelism: 2})
+	job1 := jm.CreateJobIfNotExists(&minObj1, bucket)
+	job2 := jm.CreateJobIfNotExists(&minObj2, bucket)
+	s1 := job1.subscribe(10 * util.MiB)
+	s2 := job2.subscribe(5 * util.MiB)
+	ctx := context.Background()
+
+	_, err1 := job1.Download(ctx, 10*util.MiB, false)
+	_, err2 := job2.Download(ctx, 5*util.MiB, false)
+
+	notif1, notif2 := false, false
+	timeout := time.After(1000 * time.Second)
```

Comment:
i think the timeout is too high. I think we should set it to something like 5-10s

Reply:
Sorry, it was unintended. I was using that to debug tests and missed to revert it before raising the PR.